pkg/completion/completion.go

```diff
@@ -0,0 +1,170 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package completion
+
+import (
+	"context"
+	"reflect"
+
+	"github.com/cilium/cilium/pkg/lock"
+)
+
+// Completion provides the Completed callback to be called when an
+// asynchronous computation is completed.
+type Completion interface {
+	// Context returns the context of the asynchronous computation.
+	// If the context is cancelled, e.g. if it times out, the computation must
+	// be cancelled.
+	Context() context.Context
+
+	// Completed notifies of the completion of the asynchronous computation.
+	// Idempotent. Has no effect if the Context is cancelled.
+	Completed()
+}
+
+// WaitGroup waits for a collection of Completions to complete.
+type WaitGroup interface {
+	// Context returns the context of all the Completions in the wait group.
+	Context() context.Context
+
+	// AddCompletion adds a new completion into the wait group and returns it.
+	AddCompletion() Completion
+
+	// Wait blocks until all completions added by calling AddCompletion are
+	// completed, or the context is canceled, whichever happens first.
+	// Returns the context's error if it is cancelled, nil otherwise.
+	Wait() error
+}
+
+// waitGroup waits for a collection of Completions to complete.
+type waitGroup struct {
+	// ctx is the context of all the Completions in the wait group.
+	ctx context.Context
+
+	// counterLocker locks all calls to AddCompletion and Wait, which must not
+	// be called concurrently.
+	counterLocker lock.Mutex
+
+	// pendingCompletions is the list of Completions returned by
+	// AddCompletion() which have not yet been completed.
+	pendingCompletions []Completion
+
+	// selectCases is the list of channels to select in Wait.
+	// Index 0 is the channel returned by ctx.Done().
+	// Each select case at index i > 0 corresponds to the Completion in
+	// pendingCompletions at index i-1.
+	selectCases []reflect.SelectCase
+}
+
+// NewWaitGroup returns a new WaitGroup using the given context.
+func NewWaitGroup(ctx context.Context) WaitGroup {
+	return &waitGroup{
+		ctx: ctx,
+		selectCases: []reflect.SelectCase{
+			{
+				Dir:  reflect.SelectRecv,
+				Chan: reflect.ValueOf(ctx.Done()),
+			},
+		},
+	}
+}
+
+func (wg *waitGroup) Context() context.Context {
+	return wg.ctx
+}
+
+func (wg *waitGroup) AddCompletion() Completion {
+	wg.counterLocker.Lock()
+	defer wg.counterLocker.Unlock()
+
+	done := make(chan struct{})
+	wg.selectCases = append(wg.selectCases, reflect.SelectCase{
+		Dir:  reflect.SelectRecv,
+		Chan: reflect.ValueOf(done),
+	})
+
+	c := &waitGroupCompletion{
+		ctx:  wg.ctx,
+		done: done,
+	}
+	wg.pendingCompletions = append(wg.pendingCompletions, c)
+	return c
+}
+
+func (wg *waitGroup) Wait() error {
+	wg.counterLocker.Lock()
+	defer wg.counterLocker.Unlock()
+
+	for len(wg.selectCases) > 1 {
+		chosen, _, _ := reflect.Select(wg.selectCases)
```

Comment:
I think using reflection in WaitGroups is not needed at all, `Wait` logic can be done by multiplexing completion channels into one channel and selecting from it and `wg.ctx.Done()`. Example implementation is here: https://play.golang.org/p/fHnkmHm4q8s

Please let me know if this makes sense. I wrote above snippet because I think that reflection should be used as little as possible, and I think that in this case implementation with concurrency primitives is the same in terms of complexity (and data stored in waitgroups is less complex, as we only need a channel per completion).

If you feel that this is adding unnecessary complexity to the code please dismiss review.

Reply:
@nebril Thanks for the review! I also would very much like to get rid of `reflect.Select`, especially since it's quite expensive.
However, any solution with multiple writers into a single channel makes it really hard to safely determine when to close that channel. The general rule is that a channel must be closed by its writer, since only the writer know when it stopped writing and writing into a closed channel panics. It's becoming hard when you have multiple identical writers.
I noticed that you don't close the multiplex channel in your solution.

Closing in `Wait` may work. It may look like this:
```
wg.counterLocker.Lock()
defer wg.counterLocker.Unlock()
defer close(multiplex)
for count > 0 {
    select {
    case <-multiplex:
        count--
    case <-ctx.Done():
        for _, comp := range wg.pendingCompletions {
            comp.Completed()
	}
        for count > 0 {
            <-multiplex
        }
	return wg.ctx.Err()
    }
    return nil
}
```

I'll try that.
Did I miss something?

Reply:
With that implementation, I could even get rid of the individual channel per completion, and have them write directly into the "multiplex" channel, since the implementation guarantees that we close it only after all completions are completed.

Reply:
Hmm I may need to keep individual goroutines, because the multiplex channel will remain of a bounded size, so the loop to call `comp.Completed()` may block indefinitely otherwise.
I could simply have the implementation of `Completed()` start a goroutine to write into the multiplex channel.

Reply:
Or I could simply start a goroutine for the loop itself:
```
wg.counterLocker.Lock()
defer wg.counterLocker.Unlock()
defer close(multiplex)
for count > 0 {
    select {
    case <-multiplex:
        count--
    case <-ctx.Done():
        go func() {
            for _, comp := range wg.pendingCompletions {
                comp.Completed()
	    }
        }()
        for count > 0 {
            <-multiplex
            count--
        }
	return wg.ctx.Err()
    }
    return nil
}
```

Reply:
I found a much better and simpler implementation of `Wait` without any additional channel or goroutine:
```
wg.counterLocker.Lock()
defer wg.counterLocker.Unlock()
for _, comp := range wg.pendingCompletions {
    select {
    case <-comp.done:
        continue
    case <-wg.ctx.Done():
        for _, comp := range wg.pendingCompletions {
            comp.Completed()
	}
	return wg.ctx.Err()
    }
    return nil
}
```
I can now get rid of the select cases.

Reply:
Nice!

Reply:
Done. @nebril please review the latest update.